File: mxslc/mxslc/Statements/ExpressionStatement.py
from . import Statement
from ..DataType import DataType
from ..Expressions import Expression


class ExpressionStatement(Statement):
    def __init__(self, expr: Expression):
        self.__expr = expr

    def sub_expressions(self) -> list[Expression]:
        return [self.__expr, *self.__expr.sub_expressions()]

    def instantiate_templated_types(self, template_type: DataType) -> Statement:
        expr = self.__expr.instantiate_templated_types(template_type)
        return ExpressionStatement(expr)

    def execute(self) -> None:
        self.__expr.init_evaluate()

File: mxslc++/source/statements/ExpressionStatement.cpp
//
// Created by jaket on 08/04/2026.
//

//
// Created by jaket on 28/11/2025.
//

#include "statements/ExpressionStatement.h"

#include "expressions/Expression.h"
#include "statements/interface.h"

namespace mxslc::statements
{
    ExpressionStatement::ExpressionStatement(ExprPtr expr)
        : Statement{expr->token()}, expr_{std::move(expr)}
    {

    }

    ExpressionStatement::~ExpressionStatement() = default;

    void ExpressionStatement::set_attributes(AttributeList attrs)
    {
        expr_->set_attributes(std::move(attrs));
    }

    StmtPtr ExpressionStatement::monomorphize(const TypePtr& template_type) const
    {
        ExprPtr expr = expr_->monomorphize(template_type);
        return create_statement<ExpressionStatement>(std::move(expr));
    }

    void ExpressionStatement::execute_impl() const
    {
        expr_->init();
        VarPtr _ = expr_->evaluate();
    }
}
